### DISCUSSION TITLE
Feature request: Python asyncio API

### DISCUSSION THREAD

**Discussion by mazano579** [duckdb/duckdb#9704] [Ideas]
Could we please have a Python asyncio API similar to aiosqliue? If one already exists that works please do point me to it. "aioduckdb"on pypi seems to be blocking.

**Comment by Mause** [comment]
Not sure why you think aioduckdb is blocking?

**Comment by mazano579** [comment]
Hi Elliana,

I executed the following code with "aiosqlite" and both the prroducer and the 3 consumers do work concurrently. If I swap "aioduckdb" for "aiosqlite"  the 3 consumers either don't run at all or only process one record each and stop.

Here is the code:


import asyncio
import aiosqlite


async def producer(queue: asyncio.Queue) -> None:
    # produce 10 items and put them into the queue
    for i in range(100):
        value = f"Item {i}"
        print(f"Producer: putting {value} into the queue")
        await queue.put(value)
        # sleep for a random amount of time processing data
        await asyncio.sleep(0.005)
    # indicate that the producer is done
    await queue.put(None)
    print("Producer finished")


async def consumer(queue: asyncio.Queue, n: int, database: str) -> None:
    # create a connection to the SQLite database
    async with aiosqlite.connect(database) as connection:
        # create a table to store the items
        await connection.execute("CREATE TABLE IF NOT EXISTS items (id INTEGER PRIMARY KEY, value TEXT)")
        await connection.commit()
        while True:
            # get an item from the queue
            item = await queue.get()
            # check if the item is None, which means the producer is done
            if item is None:
                # put None back into the queue for the other consumers
                await queue.put(None)
                print(f"Consumer {n} finished")
                break
            # process the item
            print(f"Consumer {n}: got {item} from the queue")
            # insert the item into the database
            print(f"Consumer {n}: inserting {item} into the database")
            await connection.execute("INSERT INTO items (value) VALUES (?)", (item,))
            await connection.commit()
            queue.task_done()


async def main():
    # create a queue
    queue = asyncio.Queue()
    # create a list of consumers
    tasks = [asyncio.create_task(producer(queue))]
    for i in range(3):
        tasks.append(asyncio.create_task(consumer(queue, i, "test.db")))
    # run the producer and wait for it to finish
    # await producer(queue)

    # await queue.join()
    # wait until all consumers are cancelled
    await asyncio.gather(*tasks, return_exceptions=True)

    # wait until the consumers have processed all the items
    print("finished all tasks")
    # await queue.join()
    # print("finished all queue")

    for task in tasks:
        task.cancel()
    print("all tasks cancelled")


if __name__ == "__main__":
    asyncio.run(main())

________________________________

**Reply by Mause** [reply]
Then you may have to write your own wrapper

**Reply by mazano579** [reply]
Thanks
